packages/runtime/container-runtime/src/test/throttler.spec.ts

```diff
@@ -134,4 +126,202 @@ describe("Throttler", () => {
+        //  8 | 2^( 7 - 1) x 20 =   64 x 20 = |  1,280 ms |  2,540 ms
+        //  9 | 2^( 8 - 1) x 20 =  128 x 20 = |  2,560 ms |  5,100 ms
+        // 10 | 2^( 9 - 1) x 20 =  256 x 20 = |  5,120 ms | 10,220 ms
+        // 11 | 2^(10 - 1) x 20 =  512 x 20 = | 10,240 ms | 20,460 ms
+        // 12 | 2^(11 - 1) x 20 = 1024 x 20 = | 20,480 ms | 40,940 ms
+        // 13 | 2^(12 - 1) x 20 = 2048 x 20 = | 30,000 ms | 70,940 ms (MAX)
+        // 14 | 2^(13 - 1) x 20 = 5096 x 20 = | 30,000 ms |100,940 ms (MAX)
+        const delayFn = formExponentialFnWithAttemptOffset(-1, {
+            coefficient: 20,
+            initialDelay: 0,
+        });
+
+        beforeEach(() => throttler = new Throttler(delayWindowMs, maxDelayMs, delayFn));
+
+        it("Should initially have zero delay", () => {
+            assert.strictEqual(throttler.getDelay(), 0);
+        });
+
+        it("Should increase as expected with instant failures", () => {
+            const expectedDelays = [
+                0, 20, 40, 80, 160, 320, 640,
+                1280, 2560, 5120, 10240, 20480,
+                30000, 30000, 30000, 30000, 30000,
+            ];
+            for (const expectedDelay of expectedDelays) {
+                assert.strictEqual(getDelayAndTick(), expectedDelay);
+            }
+        });
+
+        it("Should remain zero delay with long pauses between getDelay calls", () => {
+            for (let i = 0; i < 5; i++) {
+                assert.strictEqual(getDelayAndTick(), 0, `iteration ${i}`);
+                clock.tick(delayWindowMs);
+            }
+            assert.strictEqual(getDelayAndTick(), 0);
+
+            // This time barely keep it in the window, giving a delay.
+            clock.tick(delayWindowMs - 1);
+            assert.strictEqual(getDelayAndTick(), 20);
+        });
+
+        it("Should not increase with long pauses between getDelay calls", () => {
+            const oneThirdTicks = Math.floor(delayWindowMs / 3);
+            const remainingTicks = delayWindowMs - (2 * oneThirdTicks);
+
+            // Accumulate some attempts first.
+            assert.strictEqual(getDelayAndTick(), 0);
+            clock.tick(oneThirdTicks);
+            assert.strictEqual(getDelayAndTick(), 20);
+            clock.tick(oneThirdTicks);
+            assert.strictEqual(getDelayAndTick(), 40);
+            clock.tick(remainingTicks);
+
+            // Loop through attempts periodically dropping off.
+            for (let i = 0; i < 100; i++) {
+                assert.strictEqual(getDelayAndTick(), 40, `iteration ${i}`);
+                clock.tick(i % 3 === 2 ? remainingTicks : oneThirdTicks);
+            }
+            assert.strictEqual(getDelayAndTick(), 40);
+
+            // This time fail instantly, giving a later delay.
+            assert.strictEqual(getDelayAndTick(), 80);
+        });
+
+        it("Should stop increasing number of attempts after max", () => {
+            for (let i = 0; i < 12; i++) {
```

Comment:
perhaps `const` the value 12 with a comment about how it came to be (I assume it's num attempts within `maxDelayMs` timeframe). Same comment for line 308.

Reply:
I decided to go ahead and consolidate all the tests, let me know what you think. Now it more clearly shows the calculation for that number:
```ts
const expectedMaxAttempts = expectedDelays.length;
```

Reply:
Looks great